src/firestore/indexes.ts

```diff
@@ -31,15 +36,59 @@ export class FirestoreIndexes {
     const indexesToDeploy: Spec.Index[] = spec.indexes;
     const fieldOverridesToDeploy: Spec.FieldOverride[] = spec.fieldOverrides;
 
-    const existingIndexes = await this.listIndexes(project);
-    const existingFieldOverrides = await this.listFieldOverrides(project);
+    const existingIndexes: API.Index[] = await this.listIndexes(options.project);
+    const existingFieldOverrides: API.Field[] = await this.listFieldOverrides(options.project);
 
-    if (existingIndexes.length > indexesToDeploy.length) {
-      utils.logBullet(
-        clc.bold.cyan("firestore:") +
-          " there are some indexes defined in your project that are not present in your " +
-          "firestore indexes file. Run firebase firestore:indexes and save the result to correct the discrepancy."
-      );
+    const indexesToDelete = existingIndexes.filter((index) => {
+      return !indexesToDeploy.some((spec) => this.indexMatchesSpec(index, spec));
+    });
+
+    // We only want to delete fields where there is nothing in the local file with the same
+    // (collectionGroup, fieldPath) pair. Otherwise any differences will be resolved
+    // as part of the "PATCH" process.
+    const fieldOverridesToDelete = existingFieldOverrides.filter((field) => {
+      return !fieldOverridesToDeploy.some((spec) => {
+        const parsedName = util.parseFieldName(field.name);
+
+        if (parsedName.collectionGroupId !== spec.collectionGroup) {
+          return false;
+        }
+
+        if (parsedName.fieldPath !== spec.fieldPath) {
+          return false;
+        }
+
+        return true;
+      });
+    });
+
+    let shouldDeleteIndexes = options.force;
+    if (indexesToDelete.length > 0) {
+      if (options.nonInteractive && !options.force) {
+        utils.logLabeledBullet(
+          "firestore",
+          `there are ${indexesToDelete.length} indexes defined in your project that are not present in your ` +
+            "firestore indexes file. To delete them, run this command with the --force flag."
+        );
+      } else if (!options.force) {
+        const indexesString = indexesToDelete
+          .map((x) => this.prettyIndexString(x, false))
+          .join("\n\t");
+        utils.logLabeledBullet(
+          "firestore",
+          `The following indexes are defined in your project but are not present in your firestore indexes file:\n\t${indexesString}`
+        );
+      }
+
+      if (!options.force) {
```

Comment:
consder using `shouldDeleteIndexes` here, since that's what this block would be updating anyway